include/fir/Type.h

```diff
@@ -0,0 +1,280 @@
+public:
+  using Base::Base;
+  static FieldType get(mlir::MLIRContext *ctxt, KindTy _ = 0);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_FIELD; }
+};
+
+class HeapType : public mlir::Type::TypeBase<HeapType, mlir::Type,
+                     detail::FIRHeapTypeStorage> {
+public:
+  using Base::Base;
+  static HeapType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_HEAP; }
+
+  mlir::Type getEleTy() const;
+};
+
+class PointerType : public mlir::Type::TypeBase<PointerType, mlir::Type,
+                        detail::FIRPointerTypeStorage> {
+public:
+  using Base::Base;
+  static PointerType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_POINTER; }
+
+  mlir::Type getEleTy() const;
+};
+
+class ReferenceType : public mlir::Type::TypeBase<ReferenceType, mlir::Type,
+                          detail::FIRReferenceTypeStorage> {
+public:
+  using Base::Base;
+  static ReferenceType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_REFERENCE;
+  }
+
+  mlir::Type getEleTy() const;
+};
+
+class SequenceType : public mlir::Type::TypeBase<SequenceType, mlir::Type,
+                         detail::FIRSequenceTypeStorage> {
+public:
+  using Base::Base;
+  struct Unknown {};
+  using BoundInfo = int64_t;
+  using Extent = std::variant<Unknown, BoundInfo>;
+  using Bounds = std::vector<Extent>;
+  using Shape = std::variant<Unknown, Bounds>;
+
+  mlir::Type getEleTy() const;
+  Shape getShape() const;
+
+  static SequenceType get(const Shape &shape, mlir::Type elementType);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_SEQUENCE;
+  }
+};
+
+bool operator==(const SequenceType::Shape &, const SequenceType::Shape &);
+llvm::hash_code hash_value(const SequenceType::Extent &);
+llvm::hash_code hash_value(const SequenceType::Shape &);
+
+
+class TypeDescType : public mlir::Type::TypeBase<TypeDescType, mlir::Type,
+                         detail::FIRTypeDescTypeStorage> {
+public:
+  using Base::Base;
+  static TypeDescType get(mlir::Type ofType);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_TYPEDESC;
+  }
+  mlir::Type getOfTy() const;
+};
+
+// Derived types
+
+class RecordType : public mlir::Type::TypeBase<RecordType, mlir::Type,
+                       detail::FIRRecordTypeStorage> {
+public:
+  using Base::Base;
+  using TypePair = std::pair<std::string, mlir::Type>;
+  using TypeList = std::vector<TypePair>;
+  using LenPPair = std::pair<std::string, mlir::Type>;
+  using LenPList = std::vector<TypePair>;
+
+  llvm::StringRef getName();
+  TypeList getTypeList();
+  LenPList getLenParamList();
+
+  static RecordType get(mlir::MLIRContext *ctxt, llvm::StringRef name,
+      llvm::ArrayRef<LenPPair> lenPList = {},
+      llvm::ArrayRef<TypePair> typeList = {});
```

Comment:
What is the significance of `llvm::ArrayRef<TypePair>` versus `std::vector<TypePair>`?

Reply:
`ArrayRef` is a non-owning reference and pretty ubiquitous.

http://llvm.org/doxygen/classllvm_1_1ArrayRef.html

Reply:
We prefer to use the std types unless there's a requirement to use the llvm type for interoperability.  My understanding is that nowadays the llvm people with they had used standard types had they been available at the time.

Reply:
I don't think that is true.  MLIR is a new project and it makes use of LLVM data structures ubiquitously, for example. Why would an LLVM project intentionally avoid using LLVM data structures when they are clearly appropriate?

Reply:
It's explained here:  https://llvm.org/docs/CodingStandards.html#c-standard-library

Reply:
This reference concisely explains why LLVM projects use the LLVM ADTs.

https://llvm.org/docs/ProgrammersManual.html#picking-the-right-data-structure-for-a-task